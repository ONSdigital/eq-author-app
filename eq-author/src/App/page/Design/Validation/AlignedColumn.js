import { Column } from "components/Grid";
import styled from "styled-components";

const AlignedColumn = styled(Column)`
  align-items: center;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  height: 100%;
`;

export default AlignedColumn;
