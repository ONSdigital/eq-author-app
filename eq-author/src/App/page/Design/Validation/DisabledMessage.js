import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Message = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  flex: 1 1 auto;
  font-weight: bold;
  margin-bottom: 3em;
`;

const DisabledMessage = ({ name }) => <Message>{name} is disabled</Message>;

DisabledMessage.propTypes = {
  name: PropTypes.string.isRequired,
};

export default DisabledMessage;
