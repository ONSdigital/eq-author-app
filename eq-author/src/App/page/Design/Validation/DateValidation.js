import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { flowRight, get } from "lodash";

import { Input, Select } from "components/Forms";
import { Grid, Column } from "components/Grid";

import PreviousAnswerContentPicker from "./PreviousAnswerContentPicker";
import MetadataContentPicker from "./MetadataContentPicker";
import { ValidationPills } from "./ValidationPills";
import Path from "./path.svg?inline";
import PathEnd from "./path-end.svg?inline";
import EmphasisedText from "./EmphasisedText";
import AlignedColumn from "./AlignedColumn";
import Duration from "./Duration";

import withChangeUpdate from "enhancers/withChangeUpdate";

import * as entityTypes from "constants/validation-entity-types";
import { DATE, DATE_RANGE } from "constants/answer-types";
import { DAYS, MONTHS, YEARS } from "constants/durations";

const UNITS = [DAYS, MONTHS, YEARS];
const RELATIVE_POSITIONS = ["Before", "After"];

const DateInput = styled(Input)`
  width: 12em;
  height: 2.5em;
`;

const ConnectedPath = styled(Path)`
  height: 3.6em;
`;

const RelativePositionSelect = styled(Select)`
  width: 6em;
`;

const RelativePositionText = styled(EmphasisedText)`
  margin-top: 0.5em;
  margin-bottom: 0.5em;
`;

const StartDateText = styled.div`
  margin: 0;
  padding-top: 0.5em;
  height: 2.5em;
`;

const START_COL_SIZE = 3;
const END_COL_SIZE = 12 - START_COL_SIZE;

const getUnits = ({ format, type }) => {
  if (type === DATE_RANGE) {
    return UNITS;
  }

  if (format === "dd/mm/yyyy") {
    return UNITS;
  }

  if (format === "mm/yyyy") {
    return UNITS.slice(1);
  }

  return UNITS.slice(2);
};

export class UnwrappedDateValidation extends React.Component {
  PreviousAnswer = () => (
    <PreviousAnswerContentPicker
      answerId={this.props.answer.id}
      onSubmit={this.props.onChangeUpdate}
      selectedContentDisplayName={get(
        this.props.validation.previousAnswer,
        "displayName"
      )}
      selectedId={get(this.props.validation.previousAnswer, "id")}
      path={`answer.validation.${this.props.readKey}.availablePreviousAnswers`}
    />
  );

  Metadata = () => (
    <MetadataContentPicker
      answerId={this.props.answer.id}
      onSubmit={this.props.onChangeUpdate}
      selectedContentDisplayName={get(
        this.props.validation.metadata,
        "displayName"
      )}
      path={`answer.validation.${this.props.readKey}.availableMetadata`}
    />
  );

  Now = () => (
    <StartDateText>The date the respondent begins the survey</StartDateText>
  );

  Custom = () => (
    <DateInput
      name="customDate"
      type="date"
      value={this.props.validation.customDate}
      onChange={this.props.onChange}
      onBlur={this.props.onUpdate}
      max="9999-12-30"
      min="1000-01-01"
    />
  );

  render() {
    const {
      validation: { offset, relativePosition, entityType },
      answer: {
        properties: { format },
        type,
      },
      displayName,
      onChange,
      onUpdate,
      onChangeUpdate,
    } = this.props;

    const availableUnits = getUnits({ format, type });

    const validationPills = {
      Metadata: this.Metadata,
      Custom: this.Custom,
    };

    if (type === DATE) {
      validationPills.PreviousAnswer = this.PreviousAnswer;
      validationPills.Now = this.Now;
    }

    return (
      <div>
        <Grid>
          <AlignedColumn cols={START_COL_SIZE}>
            <EmphasisedText>{displayName} is</EmphasisedText>
          </AlignedColumn>
          <Column cols={END_COL_SIZE}>
            <Duration
              name="offset"
              duration={offset}
              units={availableUnits}
              onChange={onChange}
              onUpdate={onUpdate}
            />
          </Column>
        </Grid>
        <Grid>
          <Column cols={START_COL_SIZE}>
            <ConnectedPath />
          </Column>
        </Grid>
        <Grid>
          <AlignedColumn cols={START_COL_SIZE}>
            {type === DATE && (
              <RelativePositionSelect
                name="relativePosition"
                value={relativePosition}
                onChange={onChange}
                onBlur={onUpdate}
                data-test="relative-position-select"
              >
                {RELATIVE_POSITIONS.map(position => (
                  <option key={position} value={position}>
                    {position}
                  </option>
                ))}
              </RelativePositionSelect>
            )}
            {type === DATE_RANGE && (
              <RelativePositionText>{relativePosition}</RelativePositionText>
            )}
            <PathEnd />
          </AlignedColumn>
          <Column cols={9}>
            <ValidationPills
              entityType={entityType}
              onEntityTypeChange={onChangeUpdate}
              {...validationPills}
            />
          </Column>
        </Grid>
      </div>
    );
  }
}

UnwrappedDateValidation.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    customDate: PropTypes.string,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    metadata: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    offset: PropTypes.shape({
      unit: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    }).isRequired,
    relativePosition: PropTypes.string.isRequired,
    entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.required,
    properties: PropTypes.shape({
      format: PropTypes.string,
    }).isRequired,
  }).isRequired,
  onToggleValidationRule: PropTypes.func.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  displayName: PropTypes.string.isRequired,
  readKey: PropTypes.string.isRequired,
  testId: PropTypes.string.isRequired,
};

export default flowRight(withChangeUpdate)(UnwrappedDateValidation);
