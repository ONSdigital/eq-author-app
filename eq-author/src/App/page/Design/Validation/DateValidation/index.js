import React from "react";
import PropTypes from "prop-types";
import { flowRight, find } from "lodash";

import { Grid, Column } from "components/Grid";
import { ValidationPills } from "../ValidationPills";
import ValidationTitle from "../ValidationTitle";
import AlignedColumn from "../AlignedColumn";
import Duration from "../Duration";
import { ERR_NO_VALUE } from "constants/validationMessages";
import withChangeUpdate from "enhancers/withChangeUpdate";
import * as entityTypes from "constants/validation-entity-types";
import { DATE } from "constants/answer-types";
import { DAYS, MONTHS, YEARS } from "constants/durations";
import PathEnd from "../path-end.svg?inline";

import {
  StyledError,
  RelativePositionText,
  ConnectedPath,
  Now,
} from "./components";
import PreviousAnswerEditor from "../PreviousAnswerEditor";
import MetadataEditor from "./MetadataEditor";
import CustomEditor from "./CustomEditor";
import PositionPicker from "./PositionPicker.js";

const START_COL_SIZE = 3;
const END_COL_SIZE = 12 - START_COL_SIZE;

const getUnits = format => {
  switch (format) {
    case "dd/mm/yyyy":
      return [DAYS, MONTHS, YEARS];
    case "mm/yyyy":
      return [MONTHS, YEARS];
    default:
      return [YEARS];
  }
};

const UnwrappedDateValidation = ({
  validation,
  answer,
  displayName,
  onChange,
  onUpdate,
  onChangeUpdate,
  readKey,
}) => {
  const availableUnits = getUnits(
    answer.properties.format ? answer.properties.format : "dd/mm/yyyy"
  );

  const hasDurationError = find(validation.validationErrorInfo.errors, error =>
    error.errorCode.includes("ERR_OFFSET_NO_VALUE")
  );

  return (
    <div>
      <Grid>
        <AlignedColumn cols={START_COL_SIZE}>
          <ValidationTitle>{displayName} is</ValidationTitle>
        </AlignedColumn>
        <Column cols={END_COL_SIZE}>
          <Duration
            name="offset"
            duration={validation.offset}
            value={validation.offset.unit}
            units={availableUnits}
            onChange={onChange}
            onUpdate={onUpdate}
            hasError={hasDurationError}
          />
        </Column>
      </Grid>
      <Grid>
        <Column cols={START_COL_SIZE}>
          <ConnectedPath />
        </Column>
        <Column cols={END_COL_SIZE}>
          {hasDurationError && <StyledError>{ERR_NO_VALUE}</StyledError>}
        </Column>
      </Grid>
      <Grid>
        <AlignedColumn cols={START_COL_SIZE}>
          {answer.type === DATE ? (
            <PositionPicker
              value={validation.relativePosition}
              onChange={onChange}
              onUpdate={onUpdate}
            />
          ) : (
            <RelativePositionText>
              {validation.relativePosition.toLowerCase()}
            </RelativePositionText>
          )}
          <PathEnd />
        </AlignedColumn>
        <Column cols={9}>
          <ValidationPills
            entityType={validation.entityType}
            onEntityTypeChange={onChangeUpdate}
            Metadata={MetadataEditor}
            answer={answer}
            validation={validation}
            readKey={readKey}
            onChange={onChange}
            onUpdate={onUpdate}
            onChangeUpdate={onChangeUpdate}
            Custom={CustomEditor}
            {...(answer.type === DATE
              ? {
                  PreviousAnswer: PreviousAnswerEditor,
                  Now: Now,
                }
              : {})}
          />
        </Column>
      </Grid>
    </div>
  );
};

UnwrappedDateValidation.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    customDate: PropTypes.string,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    metadata: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    offset: PropTypes.shape({
      unit: PropTypes.string.isRequired,
      value: PropTypes.number,
    }).isRequired,
    relativePosition: PropTypes.string.isRequired,
    entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.required,
    properties: PropTypes.shape({
      format: PropTypes.string,
    }).isRequired,
  }).isRequired,
  onToggleValidationRule: PropTypes.func.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  displayName: PropTypes.string.isRequired,
  readKey: PropTypes.string.isRequired,
  testId: PropTypes.string.isRequired,
};

export default flowRight(withChangeUpdate)(UnwrappedDateValidation);
export { UnwrappedDateValidation };
