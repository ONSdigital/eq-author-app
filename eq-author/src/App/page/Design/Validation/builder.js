import { flowRight } from "lodash/fp";

import withToggleAnswerValidation from "./withToggleAnswerValidation";
import withUpdateAnswerValidation from "./withUpdateAnswerValidation";
import withEntityEditor from "components/withEntityEditor";

import withAnswerValidation from "./withAnswerValidation";
import Validation from "./Validation";
import withProps from "enhancers/withProps";
import withPropRenamed from "enhancers/withPropRenamed";
import withPropRemapped from "enhancers/withPropRemapped";

export default (displayName, testId, readKey, writeKey, readToWriteMapper) =>
  flowRight(
    withProps({ displayName, testId, readKey }),
    withAnswerValidation(readKey),
    withUpdateAnswerValidation,
    withToggleAnswerValidation,
    withPropRemapped(
      "onUpdateAnswerValidation",
      "onUpdate",
      readToWriteMapper(writeKey)
    ),
    withEntityEditor(readKey),
    withPropRenamed(readKey, "validation")
  )(Validation);
