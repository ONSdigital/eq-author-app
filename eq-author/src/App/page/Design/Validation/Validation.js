import React from "react";
import PropTypes from "prop-types";

import DisabledMessage from "./DisabledMessage";
import ValidationView from "./ValidationView";

class Validation extends React.Component {
  handleToggleChange = ({ value: enabled }) => {
    const {
      onToggleValidationRule,
      validation: { id },
    } = this.props;

    onToggleValidationRule({
      id,
      enabled,
    });
  };

  renderDisabled = () => <DisabledMessage name={this.props.displayName} />;

  render() {
    const {
      testId,
      validation: { enabled },
      children,
    } = this.props;

    return (
      <ValidationView
        data-test={testId}
        enabled={enabled}
        onToggleChange={this.handleToggleChange}
      >
        {enabled ? children(this.props) : this.renderDisabled()}
      </ValidationView>
    );
  }
}

Validation.propTypes = {
  testId: PropTypes.string,
  displayName: PropTypes.string,
  validation: PropTypes.shape({
    id: PropTypes.string,
    enabled: PropTypes.bool,
  }),
  onToggleValidationRule: PropTypes.func,
  children: PropTypes.func,
};

export default Validation;
