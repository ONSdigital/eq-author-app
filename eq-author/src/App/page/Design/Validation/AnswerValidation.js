import React from "react";
import { kebabCase, get, startCase } from "lodash";
import CustomPropTypes from "custom-prop-types";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import { gotoTab } from "redux/tabs/actions";
import ModalWithNav from "components/modals/ModalWithNav";
import SidebarButton, { Title, Detail } from "components/buttons/SidebarButton";

import ValidationContext from "./ValidationContext";
import DurationValidation from "./DurationValidation";
import DateValidation from "./DateValidation";
import NumericValidation from "./NumericValidation";
import DatePreview from "./DatePreview";
import DurationPreview from "./DurationPreview";

import {
  EarliestDate,
  LatestDate,
  MinDuration,
  MaxDuration,
  MinValue,
  MaxValue,
} from "./";

import {
  CURRENCY,
  DATE,
  DATE_RANGE,
  NUMBER,
  PERCENTAGE,
} from "constants/answer-types";

const formatValue = (value, { type }) => {
  if (typeof value !== "number") {
    return null;
  }
  if (type === PERCENTAGE) {
    return `${value}%`;
  }
  if (type === CURRENCY) {
    return `Â£${value}`;
  }
  return value;
};

export const validationTypes = [
  {
    id: "minValue",
    title: "Min Value",
    render: () => (
      <MinValue>{props => <NumericValidation {...props} />}</MinValue>
    ),
    types: [CURRENCY, NUMBER, PERCENTAGE],
    preview: ({ custom, previousAnswer, entityType }, answer) =>
      entityType === "Custom"
        ? formatValue(custom, answer)
        : get(previousAnswer, "displayName"),
  },
  {
    id: "maxValue",
    title: "Max Value",
    render: () => (
      <MaxValue>{props => <NumericValidation {...props} />}</MaxValue>
    ),
    types: [CURRENCY, NUMBER, PERCENTAGE],
    preview: ({ custom, previousAnswer, entityType }, answer) =>
      entityType === "Custom"
        ? formatValue(custom, answer)
        : get(previousAnswer, "displayName"),
  },
  {
    id: "earliestDate",
    title: "Earliest Date",
    render: () => (
      <EarliestDate>{props => <DateValidation {...props} />}</EarliestDate>
    ),
    types: [DATE, DATE_RANGE],
    preview: DatePreview,
  },
  {
    id: "latestDate",
    title: "Latest Date",
    render: () => (
      <LatestDate>{props => <DateValidation {...props} />}</LatestDate>
    ),
    types: [DATE, DATE_RANGE],
    preview: DatePreview,
  },
  {
    id: "minDuration",
    title: "Min Duration",
    render: () => (
      <MinDuration>{props => <DurationValidation {...props} />}</MinDuration>
    ),
    types: [DATE_RANGE],
    preview: DurationPreview,
  },
  {
    id: "maxDuration",
    title: "Max Duration",
    render: () => (
      <MaxDuration>{props => <DurationValidation {...props} />}</MaxDuration>
    ),
    types: [DATE_RANGE],
    preview: DurationPreview,
  },
];

const getValidationsForType = type =>
  validationTypes.filter(({ types }) => types.includes(type));

const validations = [NUMBER, CURRENCY, PERCENTAGE, DATE, DATE_RANGE].reduce(
  (hash, type) => ({
    ...hash,
    [type]: getValidationsForType(type),
  }),
  {}
);

export class UnconnectedAnswerValidation extends React.Component {
  state = {
    modalIsOpen: false,
    answerId: null,
  };

  constructor(props) {
    super(props);
    this.modalId = `modal-validation-${props.answer.id}`;
  }

  handleModalClose = () => this.setState({ modalIsOpen: false });

  renderButton = ({ id, title, value, enabled }) => (
    <SidebarButton
      key={id}
      data-test={`sidebar-button-${kebabCase(title)}`}
      onClick={() => {
        this.props.gotoTab(this.modalId, id);
        this.setState({ modalIsOpen: true });
      }}
    >
      <Title>{title}</Title>
      {enabled && value && <Detail>{value}</Detail>}
    </SidebarButton>
  );

  render() {
    const { answer } = this.props;
    const validValidationTypes = validations[answer.type] || [];
    if (validValidationTypes.length === 0) {
      return null;
    }

    return (
      <ValidationContext.Provider value={{ answer }}>
        {validValidationTypes.map(validationType => {
          const validation = get(answer, `validation.${validationType.id}`, {});
          const { enabled, previousAnswer, metadata } = validation;
          const value = enabled
            ? validationType.preview(validation, answer)
            : "";

          return this.renderButton({
            ...validationType,
            value,
            enabled,
            previousAnswer,
            metadata,
          });
        })}
        <ModalWithNav
          id={this.modalId}
          onClose={this.handleModalClose}
          navItems={validValidationTypes}
          title={`${startCase(answer.type)} validation`}
          isOpen={this.state.modalIsOpen}
        />
      </ValidationContext.Provider>
    );
  }
}

UnconnectedAnswerValidation.propTypes = {
  answer: CustomPropTypes.answer,
  gotoTab: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  tabsState: state.tabs,
});

export default connect(
  mapStateToProps,
  { gotoTab }
)(UnconnectedAnswerValidation);
