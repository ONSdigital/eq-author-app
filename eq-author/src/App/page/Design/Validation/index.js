import builder from "./builder";
import {
  dateReadToWriteMapper,
  durationReadToWriteMapper,
  numericReadToWriteMapper,
} from "./readToWriteMapper";

export const LatestDate = builder(
  "Latest date",
  "latest-date-validation",
  "latestDate",
  "latestDateInput",
  dateReadToWriteMapper
);

export const EarliestDate = builder(
  "Earliest date",
  "earliest-date-validation",
  "earliestDate",
  "earliestDateInput",
  dateReadToWriteMapper
);

export const MinDuration = builder(
  "Min duration",
  "min-duration-validation",
  "minDuration",
  "minDurationInput",
  durationReadToWriteMapper
);

export const MaxDuration = builder(
  "Max duration",
  "max-duration-validation",
  "maxDuration",
  "maxDurationInput",
  durationReadToWriteMapper
);

export const MinValue = builder(
  "Min value",
  "min-value-validation",
  "minValue",
  "minValueInput",
  numericReadToWriteMapper
);

export const MaxValue = builder(
  "Max value",
  "max-value-validation",
  "maxValue",
  "maxValueInput",
  numericReadToWriteMapper
);
