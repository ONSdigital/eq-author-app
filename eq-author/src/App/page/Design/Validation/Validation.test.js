import React from "react";
import { shallow } from "enzyme";

import Validation from "./Validation";

const createWrapper = (props, render = shallow) =>
  render(<Validation {...props} />);

describe("Validation", () => {
  let props, wrapper;
  let onToggleValidationRule = jest.fn();

  beforeEach(() => {
    props = {
      testId: "test-id",
      displayName: "display-name",
      validation: {
        id: "666",
        enabled: true,
      },
      onToggleValidationRule: onToggleValidationRule,
      children: () => <div>foobar</div>,
    };

    wrapper = createWrapper(props);
  });

  it("should render with content", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with disabled content", () => {
    props.validation.enabled = false;
    wrapper = createWrapper(props);
    expect(wrapper).toMatchSnapshot();
  });

  it("should correctly handle toggle change", () => {
    wrapper.simulate("toggleChange", { value: false });
    expect(onToggleValidationRule).toHaveBeenCalledWith({
      enabled: false,
      id: "666",
    });
  });
});
