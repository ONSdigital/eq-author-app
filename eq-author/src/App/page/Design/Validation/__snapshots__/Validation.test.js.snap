// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation should render with content 1`] = `
<ValidationView
  data-test="test-id"
  enabled={true}
  onToggleChange={[Function]}
>
  <div>
    foobar
  </div>
</ValidationView>
`;

exports[`Validation should render with disabled content 1`] = `
<ValidationView
  data-test="test-id"
  enabled={false}
  onToggleChange={[Function]}
>
  <DisabledMessage
    name="display-name"
  />
</ValidationView>
`;
